File: 0979/solution.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public int distributeCoins(TreeNode root) {
        return traverse(root)[0];
    }

    private int[] traverse(TreeNode node) {
        if (node == null) {
            return new int[] { 0, 0 };
        }
        int[] leftInfo = traverse(node.left);
        int leftMoves = leftInfo[0];
        int leftCoins = leftInfo[1];
        int[] rightInfo = traverse(node.right);
        int rightMoves = rightInfo[0];
        int rightCoins = rightInfo[1];
        int currMoves = Math.abs(leftCoins) + Math.abs(rightCoins);
        int totalMoves = leftMoves + rightMoves + currMoves;
        int currCoins = leftCoins + rightCoins + node.val - 1;
        return new int[] { totalMoves, currCoins };
    }
}

// time O(n)
// space O(n)

File: 0979/solution.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function distributeCoins(root: TreeNode | null): number {
  return traverse(root)[0];
}

function traverse(node: TreeNode | null) {
  if (node == null) {
    return [0, 0];
  }
  const [leftMoves, leftCoins] = traverse(node.left);
  const [rightMoves, rightCoins] = traverse(node.right);
  const currMoves = Math.abs(leftCoins) + Math.abs(rightCoins);
  const totalMoves = leftMoves + rightMoves + currMoves;
  const currCoins = leftCoins + rightCoins + node.val - 1;
  return [totalMoves, currCoins];
}
